phd-tests/framework/src/test_vm/config.rs

```diff
@@ -303,12 +304,40 @@ impl<'dr> VmConfig<'dr> {
             let backend_name = device_name.clone().into_backend_name();
             let device_spec = match req.interface {
                 DiskInterface::Virtio => ComponentV0::VirtioDisk(VirtioDisk {
-                    backend_name: backend_name.clone().into_string(),
+                    backend_id: SpecKey::from(
+                        backend_name.clone().into_string(),
+                    ),
                     pci_path,
                 }),
                 DiskInterface::Nvme => ComponentV0::NvmeDisk(NvmeDisk {
-                    backend_name: backend_name.clone().into_string(),
+                    backend_id: SpecKey::from(
+                        backend_name.clone().into_string(),
+                    ),
                     pci_path,
+                    serial_number: nvme_serial_from_str(
+                        device_name.as_str(),
+                        // TODO(#790): Propolis's NVMe controller used to pad
+                        // serial numbers with 0 bytes by default. This causes
+                        // disk serial numbers to appear in the guest to be
+                        // null-terminated strings. This, in turn, affects (or
+                        // at least may affect) how guest firmware images
+                        // identify disks when determining a VM's boot order: a
+                        // disk whose serial number is padded to 20 bytes with
+                        // spaces may not match a disk whose serial number was
+                        // padded with \0 bytes.
+                        //
+                        // Unfortunately for us, guest firmware may persist disk
+                        // identifiers to a VM's nonvolatile EFI variable store.
+                        // This means that changing from zero-padded to
+                        // space-padded serial numbers may break an existing
+                        // VM's saved boot order.
+                        //
+                        // Until we decide whether and how to preserve
+                        // compatibility for existing VMs that may have
+                        // preserved a zero-padded disk ID, continue to zero-pad
+                        // disk IDs in PHD to match previous behavior.
```

Comment:
i initially wondered if you expected to copy this comment to Omicron later, but the corresponding Omicron change will probably be a migration to persist the null-padded serials and switch to space-padded for new disks huh? then at least any old wrong data will be made static, rather than "fixed" without us expecting.

Reply:
Cleaned up in d6dd7c7. The main thing I wanted to emphasize in this comment is that PHD behaves this way because Omicron will when I create the initial changes to sled-agent to get it to use the spec-based ensure API), and I don't want their behaviors to drift too far apart.

> the corresponding Omicron change will probably be a migration to persist the null-padded serials and switch to space-padded for new disks huh?

This is definitely the most straightforward approach, but I don't really *like* it, because it entails having Nexus have to keep track of the (hopefully small-and-getting-smaller) set of existing disks that need the special "pad with zeroes not spaces" compat quirk. I would be willing to consider some possibly-outlandish options (e.g. trying to do something clever in the boot order management logic in our EDK2 fork) to avoid adding a quirk column/table just for this.

It probably goes without saying, but I intend not to try to solve that problem in the Omicron change that will ingest these Propolis changes--the first order of business is to convert to the new API without actually changing or breaking anything :)